timeutil: refresh timer take 2

bboozzoo opened the pull request:
Another take on the new refresh timer parser. I adopted the structure suggested by @niemeyer in #4269.

zyga commented:
Just looking at the diffstat and the commit log, can this be split into "add new parser data types", "add new parser logic", "switch to new parser", "drop old parser". Each should iterate faster and influence the rest.

bboozzoo commented:
Hmm, I'm somewhat unconvinced, but let's try this anyway. I'll push force a split patchset.

bboozzoo commented:
Split into smaller pieces and force pushed. I tried to keep the pieces that are related inside the same commits. Hopefully the changes did not get mixed too much.

bboozzoo commented:
FWIW, while exercising, I thought of a test case that the current code will fail. I'll push a fix tomorrow.

mvo5 commented on timeutil/schedule.go:
```diff
@@ -82,40 +108,285 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+	if w.Pos == 0 {
+		return w.Day
+	}
+	return w.Day + strconv.FormatUint(uint64(w.Pos), 10)
```
(nitpick) you could use `strconv.Itoa(int(w.Pos))` instead of FormatUint() which is slightly shorter.

mvo5 commented on timeutil/schedule.go:
```diff
@@ -233,3 +514,302 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+
+var specialTokens = map[string]string{
+	// shorthand variant of whole day
+	"-": "0:00-24:00",
```
Are those special tokens tested?

bboozzoo replied:
There are tests for them right here: https://github.com/snapcore/snapd/pull/4313/commits/92325cfada9f6c53bd6eae917eb8f4e6f56a670f#diff-9a58566663280b9c066e815135f62b64R476

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +514,302 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+// mon..wed (same as above)
+//
+// Returns a slice of schedules or an error if parsing failed
+func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
```
I haven't looked into the parser. Let's sort out the rest first and then I'll dive in here.

bboozzoo replied:
I'd appreciate if you could look at the parser code anyway.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -82,40 +108,285 @@ func ParseTime(s string) (t TimeOfDay, err error) {
-func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
+func (sched *Schedule) weekSpans() []WeekSpan {
+	if len(sched.Week) == 0 {
+		return []WeekSpan{{}}
```
Alternatively, if the algorithm looks simpler with this having a default span, we need to be explicit and cook a proper span instead of trusting zero values.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -67,6 +88,11 @@ func IsValidWeekday(s string) bool {
 // an TimeOfDay type or an error
 func ParseTime(s string) (t TimeOfDay, err error) {
+	if s == "24:00" {
+		t.Hour = 24
```
It's needed to indicate that the time span covers the whole day 0:00-24:00, 24:00 being the other end of the day. Given this, 0:00-0:00 is a single even at midnight, 0:00-24:00 is an event from midnight until the next midnight.

niemeyer replied:
Good point!

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -57,6 +61,23 @@ func (t TimeOfDay) Add(dur time.Duration) TimeOfDay {
 
+// MakeTime generates a time.Time using base for information on year, month, day
+// and with hour and minute set from TimeOfDay
+func (t TimeOfDay) MakeTime(base time.Time) time.Time {
```
`timeutil.Clock` sounds good to me

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -82,40 +108,285 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+// week-in-the-month position, eg. the first Monday of the month
+type Weekday struct {
+	// Day stands for a weekday - "mon", "tue", ..
+	Day string
```
Good point. I used string here so that I could tell if it's unset and then such Weekday becomes a match-all. With your other suggestions, there will be no fallback, empty WeekSpans to match against, so switching to time.Weekday should be just fine.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -82,40 +108,285 @@ func ParseTime(s string) (t TimeOfDay, err error) {
-			continue
+		t = tnext
+
+		tnext = t.Add(24 * time.Hour)
```
There are tests for this https://github.com/snapcore/snapd/pull/4313/files#diff-9a58566663280b9c066e815135f62b64R730. Note that the `tnext` which modified here is assigned to `t` only on the next iteration. So, before `t` jumps 24 hours to the next day, all time spans will be considered for the current day.

niemeyer replied:
Sorry, you are right and I was confused.

Can we please do something more orthodox:
```
for t := last; ; t = t.Add(24h) {

bboozzoo replied:
Yup, looks much clearer. Thanks for the suggestion.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -57,6 +61,23 @@ func (t TimeOfDay) Add(dur time.Duration) TimeOfDay {
+}
+
+// isValidTime returns true if given s looks like a valid time specification
+func isValidTime(s string) bool {
```
Had to refresh my memory, but I pondered on @stolowski's suggestion before and I did not manage to come up with a reasonable regexp that would not result in having more match groups. For instance ``^(([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9]))|(24:00)$`` puts 24:00 in separate group. The upside is that we can use `.MatchString()` for simple validation, the dowside though is that `24:00` becomes a special case returned in a separate group. At the end of the day, maybe that extra group is not a big problem after all.

bboozzoo commented:
Phew, lots of comments. Appreciate the review. I'll start addressing them with proper fixes on Monday.

bboozzoo commented:
Pushed fixes. I tried to put each 'fix' in a small, self contained patch.  I hope to have addressed all of the comments. The code looks much cleaner now, thanks for all suggestions.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -28,35 +29,49 @@ import (
 	return fmt.Sprintf("%02d:%02d", t.Hour, t.Minute)
 }
 
 // Sub subtracts `other` TimeOfDay from current and returns duration
```
There are multiple occurrences of TimeOfDay still. Please grep throughout the whole code base to make sure we've moved it all over.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -28,35 +29,49 @@ import (
+}
+
+// isValidTime returns true if given s looks like a valid time specification
+func isValidTime(s string) bool {
```
This function can now go away, as the short oneliner below is nice on its own and is only ever used on a single other line.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -66,56 +81,307 @@ func IsValidWeekday(s string) bool {
 // ParseTime parses a string that contains hour:minute and returns
 // an TimeOfDay type or an error
-func ParseTime(s string) (t TimeOfDay, err error) {
+func ParseTime(s string) (t Clock, err error) {
```
This may now be ParseClock, etc. Not sure if there is something else..

bboozzoo commented:
Pushed more fixes and meged master to fix the conflicts.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	for _, event := range events {
+		// process events one by one
+
+		randomized := false
```
_var randomized bool_

Fits better with what follows.

niemeyer replied:
Except, this should be spread I believe.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	var expectTime bool
+
+	for _, event := range events {
+		// process events one by one
```
I think that's sort of implied by the former line. :-)  &lt;/snark&gt;

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+		} else if c > 0 {
+			// count was specified
+			count = c
+			// count token is only allowed in timespans, meaning
```
I'll stop reviewing the logic here as the suggestion above will significantly change this, hopefully simplifying it. We can tell we only expect times because we got the first time, and we can tell we got the first time because it contains a : and we parsed it. The logic here feels a bit too complex for what is going on. I'd expect more like split, parseWeekSpan, parseWeekSpan, parseTimeSpan, parseTimeSpan, over.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -137,66 +398,69 @@ var (
 // and extracts the start and end of that schedule string and
 // returns them and any errors.
-func parseTimeInterval(s string) (start, end TimeOfDay, err error) {
+func parseTimeInterval(s string) (start, end Clock, err error) {
```
Please ignore this.. see latter comments.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+}
+
+// parseSpan splits a span string `<start>-<end>` and returns the start and end
+// pieces. If string is not a span then the whole string is returned
```
Please ignore this until you read the rest of the comments.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -137,66 +398,69 @@ var (
-	"fri": 5,
-	"sat": 6,
+var weekdayMap = map[string]time.Weekday{
+	"sun": time.Sunday,
```
`time.Sunday` is actually 0

niemeyer replied:
Ah, sorry about the silly comment then. That's exactly what I was looking for.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -28,94 +29,354 @@ import (
+func (sched *Schedule) flattenedTimeSpans() []TimeSpan {
+	baseTimes := sched.TimeSpans
+	if len(baseTimes) == 0 {
+		baseTimes = []TimeSpan{{}}
```
In order of preference

>  1) Force a time to be specified;

+1 for this

> 3) Define a default schedule window eg. 0:00-6:00

+0.5

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+	var schedule []*Schedule
+
+	for _, s := range strings.Split(scheduleSpec, "..") {
```
Is this what you mean?
```
mon,10:00_fri,15:00 (Monday at 10:00, Friday at 15:00)
mon,fri,10:00,15:00 (Monday at 10:00 and 15:00, Friday at 10:00 and 15:00)
mon-wed,fri,9:00-11:00/2 (Monday to Wednesday and on Friday, twice between
                          9:00 and 11:00)
mon,9:00~11:00_wed,22:00~23:00 (Monday, sometime between 9:00 and 11:00, and
                                on Wednesday, sometime between 22:00 and 23:00)
mon,wed  (Monday and on Wednesday)
mon_wed  (same as above)
```
There's only few non-shell-interfering characters left that we can use. Other variants:
```
mon,9:00~11:00+wed,22:00~23:00
mon,9:00~11:00/wed,22:00~23:00
```

niemeyer replied:
Yes, we're indeed short on good options given shell special characters.

The problem with these options above as schedule separators is:

- Underline ("_") connects the words too much (think "mon_wed,11:00")
- Slash ("/") is already used for splits
- Space (" ") needs quoting
- Plus ("+") may be interpreted as a timezone (think "11:00+02:00")

Here is another idea, which has the benefit of not introducing additional special characters, and thus makes it look more even:

`mon,9:00~11:00,,wed,22:00~23:00`

stolowski commented on timeutil/schedule.go:
```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+
+// Includes returns whether t falls inside the window.
+func (s ScheduleWindow) Includes(t time.Time) bool {
+	return (t.Equal(s.Start) || t.After(s.Start)) && (t.Equal(s.End) || t.Before(s.End))
```
more compact - check if we're outside of the start-end range: `return !(t.Before(s.Start) || t.After(s.End))`

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	}
+
+	if (parsed.Start.Pos != 0) != (parsed.End.Pos != 0) {
+		return span, fmt.Errorf("cannot parse %q: mixed weekday and nonweekday", s)
```
We'd not get here if these were not weekdays, after all they were parsed successfully. So perhaps:

"cannot parse %q: week number must be present in both or neither"

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+		parsed.Spread = true
+		rest = strings.Replace(rest,
+			spreadToken,
+			spanToken, 1)
```
This should be a single line. We don't follow the 80 columns rule for Go in general, and this one in particular feels a bit over cramped.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	}
+
+	if strings.Contains(rest, spanToken) {
+		if rest == "-" {
```
This sequence can be simplified as:
```
if rest == "-" {
    ...
} else if strings.Contains(...) {
    ...
} else {
    ...
}
```

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+		}
+	} else {
+		parsed.Start, err = ParseClock(rest)
+		if err == nil {
```
No need to test here. This is a bit more relevant considering we're handling the error later. Makes more obvious that it's not _just_ about conditionally setting End, but rather that err is being looked at elsewhere.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+// special tokens `-`, `~` (followed by an optional [/count]) that indicate a
+// whole day span, or a whole day span with spread respectively.
+func parseClockSpan(s string) (span ClockSpan, err error) {
+	var parsed ClockSpan
```
Let's please use span instead of parsed. This is pretty much the only reason to have the the variables named, since in this particular case the types are enough of a hint for the person reading the function to know what these mean.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+
+	parsed.Split, rest, err = parseCount(s)
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: not a valid interval", s)
```
And here it's safer to return ClockSpan{}. It's a bit more verbose, but also safer and more clear that we're not interested in a partial result.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	}
+
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: not a valid time", s)
```
Same here about span.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+}
+
+// parseCount will parse the string containing a count token and return the
+// count, the rest of the string with count information removed or an error
```
// ...
// count and the rest of the string with count information removed, or an error.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+			return nil, fmt.Errorf("cannot parse %q: not a valid event", s)
+		}
+
+		if strings.Contains(event, ":") || shorthandTime[event[0:1]] {
```
`... || strings.ContainsAny(event[:1], "-~")`

shorthandTime may be dropped.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	"~": true,
+}
+
+// Parse each event and return a slice of schedules.
```
Not returning a slice, also "each event" doesn't match anything meaningful given the prototype.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+// Parse each event and return a slice of schedules.
+func parseEventSet(s string) (*Schedule, error) {
+	var events []string
+	// split event set into events
```
That's not conceptually right. The individual parts of this string are not events. They are weekdays, or weekday ranges, or times, or time ranges. None of these are events. The event is something like "wednesday at 11:00", which can't be extrapolated from any individual entry in that list. Those are just the forming pieces via which we compute the event set.

We need to rename this to something like "parts" or "fragments"

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	//     timelist = timeset *( "," timeset )
+	//
+	// NOTE: the syntax is ambiguous in the sense the type of a 'set' is now
+	// explicitly indicated, events with : inside are expected to be
```
Same here. Please revise the terminology and comments to make sure we're not referring to these pieces as events.

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+
+	for _, event := range events {
+		if len(event) == 0 {
+			return nil, fmt.Errorf("cannot parse %q: not a valid event", s)
```
Same!

niemeyer commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+			schedule.WeekSpans = append(schedule.WeekSpans, span)
+		} else {
+			// not a timeset
+			return nil, fmt.Errorf("cannot parse %q: expected time spec", event)
```
These pops up to users, so let's not refer to "specs" or similar coding slang:

"cannot parse %q: invalid schedule fragment"

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	countToken  = "/"
+)
+
+var shorthandTime = map[string]bool{
```
As discussed during the meeting, the shorthand syntax will be removed.

bboozzoo commented on timeutil/schedule.go:
```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
 
-func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
-	now := timeNow()
+func (sched *Schedule) flattenedTimeSpans() []ClockSpan {
```
Made this one `flattenedClockSpans`, as there's already a field named `Schedule.ClockSpans`

niemeyer replied:
That's what I meant too, sorry for being too vague.

zyga commented on timeutil/schedule.go:
```diff
@@ -28,49 +29,57 @@ import (
 	}
+
+	if m[0] == "24:00" {
+		t.Hour = 24
```
Don't you need to set t.Minute as well? Also why is this not just handled generally below?

bboozzoo replied:
`t.Minutes` will default to 0 which is just fine in this case.

zyga commented on timeutil/schedule.go:
```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+	// Pos defines which week inside the month the Day refers to, where zero
+	// means every week, 1 means first occurrence of the weekday, and 5
+	// means last occurrence (which might be the fourth or the fifth).
+	Pos uint
```
Nitpick:

```
typedef WeekOfMonth uint
const (
  EveryWeek WeekOfMonth = iota,
  FirstWeekOfMonth,
  SecondWeekOfMonth,
  ThirdWeekOfMonth,
  FourthWeekOfMonth,
  LastWeekOfMonth,
)
```

bboozzoo replied:
IIRC we've gone back and fort a bit on this. I can try and add:
```
const (
   EveryWeek uint = 0
   LastWeek uint = 5
)
```

niemeyer replied:
What we have now seems better. The rationale is that the numbers is what people will see (except for zero, which is the default). It's awkward to map the number which is publicly visible into type names. It does the opposite of what a constant name is generally supposed to do. mon2 is monday, the second. We don't need a different name for 2.

zyga commented on timeutil/schedule.go:
```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+func (w Week) String() string {
+	// Wednesday -> wed
+	day := strings.ToLower(w.Weekday.String()[0:3])
+	if w.Pos == 0 {
```
```
if w.Pos == EveryWeek {
```

zyga commented on timeutil/schedule.go:
```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+		// is it the right week?
+		week := uint(t.Day()/7) + 1
+
+		if start.Pos == 5 {
```
```
if start.Pos == LastWeekOfMonth {
```

zyga commented on timeutil/schedule.go:
```diff
@@ -28,49 +29,57 @@ import (
 	"time"
 )
 
```
I'm not sure if you agree but in reading this it helped me to put the 
vocabulary/context in my head:

```
// This module contains a number of types, here's a quick guide:
//
// Clock - tuple(hour, minute)
//   Just the time of day with minute accuracy
// ClockSpan - tuple(start, end, split, spread)
//   ClockSpan takes two Clock values to offer a `start`-`stop` range
//   and additionally allows to partition it into up to `split` chunks
//   that can be randomized if `spread` is on.
// Week - tuple(weekeday, week-of-month)
//   Week combines the actual weekday, like Monday, with an optional
//   number of the week in a month, e.g. last week of month.
// WeekSpan - tuple(start, end), pair of Week types
//   WeekSpan is like a range inside some month, but not across months.
// Schedule - tuple(week-spans, clock-spans)
//   Schedule ties everything together allowing to set a number of
//   week-spans and clock-spans, applying a NxM combination of all of them.
```

niemeyer replied:
That looks like documenting every type again.

Note that godoc works locally too, and can produce textual output into the terminal.

zyga commented on timeutil/schedule.go:
```diff
@@ -224,12 +449,243 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+	}
+
+	if (parsed.Start.Pos != 0) != (parsed.End.Pos != 0) {
+		return span, fmt.Errorf("cannot parse %q: week number must be present in both or neither", s)
```
In both or neither ... what? It feels like a word is missing.
